### PR TITLE
use c/image/pkg/blobcache

### PR DESCRIPTION
**Pull request by vrothberg**
Buildah's pkg/blobcache has been moved into containers/image to
consolidate implementations of interfaces such as image destinations
and image sources.

Create a type alias in pkg/buildcache and redirect NewBlobCache to the
containers/image function.  While it is still an API break as it changes
an interface to a struct, known callers will continue building since
they are not implementing the interface.

Since there are no functional changes: [NO NEW TESTS NEEDED]

[1] https://github.com/containers/image/pull/1459

Signed-off-by: Valentin Rothberg <vrothberg@redhat.com>

### CODE REVIEW

**Comment by nalind** [comment]
Any reason to not wait until it's in an actually-tagged release of the image library?  Please add a type alias and `NewBlobCache()` wrapper.

**Comment by vrothberg** [comment]
> Any reason to not wait until it's in an actually-tagged release of the image library?

My memory is still fresh on it, and I fear we'd forget about it.

> Please add a type alias and `NewBlobCache()` wrapper.

Note that the BlobCache in c/image is a struct, not an interface. It should compile just fine but strictly speaking it's an API break. Still cool with it?

**Comment by nalind** [comment]
Dependabot will catch it.  Does the struct not implement the interface?

File: push.go [hunk 1]
```diff
@@ -21,6 +21,21 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+// cacheLookupReferenceFunc wraps a BlobCache into a
+// libimage.LookupReferenceFunc to allow for using a BlobCache during
+// image-copy operations.
+func cacheLookupReferenceFunc(directory string, compress types.LayerCompression) libimage.LookupReferenceFunc {
+	// NOTE: this prevents us from moving BlobCache around and generalizes
+	// the libimage API.
+	return func(ref types.ImageReference) (types.ImageReference, error) {
```
[the remark this answers is not in the document]
Comment (mtrmac) [reply push.go:30]: > Return `ref` if `directory` is empty.

BTW that could simplify the callers.

File: push.go [hunk 2]
```diff
@@ -21,6 +21,21 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+// cacheLookupReferenceFunc wraps a BlobCache into a
+// libimage.LookupReferenceFunc to allow for using a BlobCache during
+// image-copy operations.
+func cacheLookupReferenceFunc(directory string, compress types.LayerCompression) libimage.LookupReferenceFunc {
+	// NOTE: this prevents us from moving BlobCache around and generalizes
+	// the libimage API.
```
Comment (mtrmac) [inline push.go:29]:
```suggestion
	// Using a closure here allows us to reference a BlobCache without having to explicitly
	// maintain it in the libimage API.
```
… but even that feels more like a comment for the  `libimageOptions.{Source,Destination}LookupReferenceFunc` field definition (as a rationale for using that design), it’s not specific to this one use.

**Comment by mtrmac** [comment]
> Please add a type alias and `NewBlobCache()` wrapper.

(Thinking about the `ClearCache` method, and to minimize round-trips) Is the goal to keep the Go-visible API surface stable (i.e. caring about the struct/interface distinction), as a matter of policy/principle (similar to c/image), or to keep a specific external consumer working (without caring about the struct/interface distinction)?

Should the existing Buildah callers use Buildah’s wrapper, or call the c/image version directly?

----

> The `ClearCache` was made private but I think we can make it public again. @mtrmac are you cool with that?

Yes; I can’t see a strong reason not to expose that.

**Comment by nalind** [comment]
We're a library, so it's supposed to be a principle that we compromise on only rarely and with care.  Historically, not everyone agrees with that, and I make mistakes, too.

**Comment by mtrmac** [comment]
> We're a library, so it's supposed to be a principle that we compromise on only rarely and with care.

Makes sense, thanks.

**Comment by vrothberg** [comment]
@nalind @mtrmac PTAL

**Review by mtrmac — COMMENTED** [review]
Implementation mechanism LGTM. I understood the discussion to mean that Buildah’s type should remain an interface — up to @nalind .

**Comment by vrothberg** [comment]
> Implementation mechanism LGTM. I understood the discussion to mean that Buildah’s type should remain an interface — up to @nalind .

Let's do that.

**Comment by TomSweeneyRedHat** [comment]
LGTM
but definitely want a @nalind head nod

**Comment by nalind** [comment]
Implementation looks fine, but again, I'd rather we wait until this is in an actual release of the image library.

**Comment by mtrmac** [comment]
> I'd rather we wait until this is in an actual release of the image library.

Let’s do that: https://github.com/containers/image/pull/1482 .

**Comment by vrothberg** [comment]
> @vrothberg c/image 5.20.0 has been released.

Thanks for taking care!

**Pull request merged**